ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/PivotBlockConfirmer.java

```diff
@@ -0,0 +1,232 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * This task will query {@code numberOfPeersToQuery} peers for a particular block number. If any
+ * peers disagree on the block at this number, the task fails with a {@code
+ * ContestedPivotBlockException}. The task will succeed only if {@code numberOfPeersToQuery}
+ * distinct peers all return matching block headers for the specified block number.
+ *
+ * @param <C> The consensus context
+ */
+class PivotBlockConfirmer<C> {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final EthContext ethContext;
+  private final MetricsSystem metricsSystem;
+  private final ProtocolSchedule<C> protocolSchedule;
+
+  // The number of peers we need to query to confirm our pivot block
+  private final int numberOfPeersToQuery;
+  // The current pivot block number, gets pushed back if peers disagree on the pivot block
+  private final long pivotBlockNumber;
+  // The number of times to retry if a peer fails to return an answer to our query
+  private final int numberOfRetriesPerPeer;
+
+  private final CompletableFuture<FastSyncState> result = new CompletableFuture<>();
+  private final Collection<CompletableFuture<?>> runningQueries = new ConcurrentLinkedQueue<>();
+  private final Map<BytesValue, RetryingGetHeaderFromPeerByNumberTask> pivotBlockQueriesByPeerId =
+      new ConcurrentHashMap<>();
+  private final Map<BlockHeader, AtomicInteger> pivotBlockVotes = new ConcurrentHashMap<>();
+
+  private final AtomicBoolean isStarted = new AtomicBoolean(false);
+  private final AtomicBoolean isCancelled = new AtomicBoolean(false);
+
+  PivotBlockConfirmer(
+      final ProtocolSchedule<C> protocolSchedule,
+      final EthContext ethContext,
+      final MetricsSystem metricsSystem,
+      final long pivotBlockNumber,
+      final int numberOfPeersToQuery,
+      final int numberOfRetriesPerPeer) {
+    this.protocolSchedule = protocolSchedule;
+    this.ethContext = ethContext;
+    this.metricsSystem = metricsSystem;
+
+    this.pivotBlockNumber = pivotBlockNumber;
+    this.numberOfPeersToQuery = numberOfPeersToQuery;
+    this.numberOfRetriesPerPeer = numberOfRetriesPerPeer;
+  }
+
+  public CompletableFuture<FastSyncState> confirmPivotBlock() {
+    if (isStarted.compareAndSet(false, true)) {
+      LOG.info(
+          "Confirm pivot block {} with at least {} peers.", pivotBlockNumber, numberOfPeersToQuery);
+      queryPeers(pivotBlockNumber);
+    }
+
+    return result;
+  }
+
+  private void queryPeers(final long blockNumber) {
+    synchronized (runningQueries) {
+      for (int i = 0; i < numberOfPeersToQuery; i++) {
+        final CompletableFuture<?> query =
```

Comment:
suggestion: consider not using a local variable

Reply:
I think it's a little more readable this way.  Why remove the local variable?